base_layer/core/src/block.rs

```diff
@@ -51,29 +65,69 @@ impl Block {
     /// This function will check the block to ensure that all UTXO's are validly constructed and that all signatures are
     /// valid. It does _not_ check that the inputs exist in the current UTXO set;
     /// nor does it check that the PoW is the largest accumulated PoW value.
-    pub fn check_internal_consistency(&self) -> Result<(), TransactionError> {
-        let mut trans: Transaction = self.body.clone().into(); // todo revisit this one q=whole code chain is completed
+    pub fn check_internal_consistency(&self) -> Result<(), BlockError> {
+        self.check_kernel_rules()?;
+        self.check_spent_utxo_rules()?;
+        let mut trans: Transaction = self.body.clone().into();
         trans.offset = self.header.total_kernel_offset.clone();
+        // We need to add the coinbase off set to the block otherwise it will not balance.
+        self.add_coinbase_offset(&mut trans);
+        // no lets check the block
         trans.validate_internal_consistency(&PROVER, &COMMITMENT_FACTORY)?;
         self.check_pow()
     }
 
-    pub fn check_pow(&self) -> Result<(), TransactionError> {
+    /// This function will check spent utxo rules like coinbase maturity height etc
+    pub fn check_spent_utxo_rules(&self) -> Result<(), BlockError> {
+        // for txo in &self.body.inputs{
+        //     if txo.features.contains(OutputFeatures::COINBASE_OUPUT){
+        // ToDo fill this out in a seperate PR.  as this requires extra code and modifications to the MMR,
+        // we currently cannot get the creation height of a given  commitment.
+        //     }
+        // }
         Ok(())
     }
 
+    /// This function will check spent kernel rules like tx lock height etc
+    pub fn check_kernel_rules(&self) -> Result<(), BlockError> {
+        for kernel in &self.body.kernels {
+            if kernel.lock_height > self.header.height {
+                return Err(BlockError::InvalidKernel);
+            }
+        }
+        Ok(())
+    }
+
+    /// This function will check that the supplied PoW is valid
+    pub fn check_pow(&self) -> Result<(), BlockError> {
+        self.header
+            .pow
+            .validate_pow(&self.body)
```

Comment:
Why not use a `?` operator

Reply:
the error types differ there. I use the map error command to return the correct error